### PR TITLE
CSU-10 Add QPU information endpoints to ZMQ server

### PR DESCRIPTION
**Pull request by jfriel-oqc**
For completion, pending hardware model changes for full error set and all QPU properties 
See this issue for current design https://github.com/oqc-community/qat/issues/127
currently being worked on

### CODE REVIEW

**Comment by jfriel-oqc** [comment]
I have read the Contributor License Agreement and I hereby accept the Terms.

**Comment by jfriel-oqc** [comment]
recheck

File: src/QAT_RPC/qat_rpc/zmq/wrappers.py [hunk 1]
```diff
@@ -1,13 +1,23 @@
...
 
 import zmq
 from qat.purr.backends.echo import get_default_echo_hardware
 from qat.purr.compiler.config import CompilerConfig
 from qat.purr.compiler.hardware_models import QuantumHardwareModel
 from qat.purr.compiler.runtime import get_runtime
+from qat.purr.integrations.features import OpenPulseFeatures
 from qat.qat import execute_with_metrics
 
+class Messages(Enum):
```
Comment (keriksson-rosenqvist-oqc) [inline src/QAT_RPC/qat_rpc/zmq/wrappers.py:17]: I think some of this should live in a `qat_rpc/utils.py` module or even a central `Server` class, as it's not ZMQ specific, and there's potentially going to be other rpc methods in the future.
OK for now though.

File: src/QAT_RPC/qat_rpc/zmq/wrappers.py [hunk 2]
```diff
@@ -65,17 +75,56 @@ def __init__(self, hardware: QuantumHardwareModel = None):
...
+        raise NotImplementedError("Individual qubit information not implented, pending hardware model changes.")
+
+    def _qpu_info(self):
+        features = OpenPulseFeatures()
+        features.for_hardware(self._hardware)
+        qpu_info = features.to_json_dict()
+        return {"qpu_info": qpu_info}
+    
+    def _interpret_message(self, message):
+        match message[0]:
```
Comment (keriksson-rosenqvist-oqc) [inline src/QAT_RPC/qat_rpc/zmq/wrappers.py:114]: This syntax is not supported in python < 3.10

**Review by keriksson-rosenqvist-oqc — DISMISSED** [review]
Generally happy, only need to adjust for comparability with older python versions.

File: src/QAT_RPC/qat_rpc/zmq/wrappers.py (hunk 2 above)
Comment (bgsach) [reply src/QAT_RPC/qat_rpc/zmq/wrappers.py:114]: Looks like we are accepting qat-rpc is now python 3.10+ ?

**Comment by bgsach** [comment]
I have read the Contributor License Agreement and I hereby accept the Terms.

**Comment by bgsach** [comment]
This is blocked waiting for the next QAT release which fixes the test failure seen above. The bug was actually in QAT not QAT-RPC.

See https://github.com/oqc-community/qat/pull/221

**Review by bgsach — APPROVED** [review]
LGTM subject to updates.

File: src/tests/zmq/test_wrappers.py [hunk 3]
```diff
@@ -1,17 +1,28 @@
...
+    add_direction_couplings_to_hardware,
+    get_default_echo_hardware,
+)
 
 from qat_rpc.utils.constants import PROMETHEUS_PORT
 from qat_rpc.utils.metrics import MetricExporter, PrometheusReceiver
 from qat_rpc.zmq.wrappers import ZMQClient, ZMQServer
 
+qubit_count = 8
+qpu_couplings = [(i, j) for i in range(qubit_count) for j in range(qubit_count) if i != j]
```
Comment (bgsach) [inline src/tests/zmq/test_wrappers.py:16]: There was previously no check that i!=j which caused flakeyness in the tests.

**Review by keriksson-rosenqvist-oqc — APPROVED** [review]
LGTM once CI passes

**Pull request merged**